File: client/exports.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package client

var (
	calls []func(sc *ScCallContext)
	views []func(sc *ScCallContext)
)

//export sc_call_entrypoint
func scCallEntrypoint(index int32) {
	if (index & 0x8000) != 0 {
		views[index&0x7fff](&rootViewContext)
		return
	}
	calls[index](&rootCallContext)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScExports struct {
	exports ScMutableStringArray
}

func NewScExports() ScExports {
	root := ScMutableMap{objId: 1}
	return ScExports{exports: root.GetStringArray("exports")}
}

func (ctx ScExports) AddCall(name string, f func(sc *ScCallContext)) {
	index := int32(len(calls))
	calls = append(calls, f)
	ctx.exports.GetString(index).SetValue(name)
}

func (ctx ScExports) AddView(name string, f func(sc *ScCallContext)) {
	index := int32(len(views))
	views = append(views, f)
	ctx.exports.GetString(index | 0x8000).SetValue(name)
}

func Nothing(sc *ScCallContext) {
	sc.Log("Doing nothing as requested. Oh, wait...")
}

File: rust/wasplib/src/client/exports.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// encapsulates standard host entities into a simple interface

use super::context::*;
use super::mutable::*;

static mut CALLS: Vec<fn(&ScCallContext)> = vec![];
static mut VIEWS: Vec<fn(&ScViewContext)> = vec![];

#[no_mangle]
fn sc_call_entrypoint(index: i32) {
    unsafe {
        if (index & 0x8000) != 0 {
            VIEWS[(index & 0x7fff) as usize](&ROOT_VIEW_CONTEXT);
            return;
        }

        CALLS[index as usize](&ROOT_CALL_CONTEXT);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScExports {
    exports: ScMutableStringArray,
}

impl ScExports {
    pub fn new() -> ScExports {
        let root = ScMutableMap::new(1);
        ScExports { exports: root.get_string_array("exports") }
    }

    pub fn add_call(&self, name: &str, f: fn(&ScCallContext)) {
        unsafe {
            let index = CALLS.len() as i32;
            CALLS.push(f);
            self.exports.get_string(index).set_value(name);
        }
    }

    pub fn add_view(&self, name: &str, f: fn(&ScViewContext)) {
        unsafe {
            let index = VIEWS.len() as i32;
            VIEWS.push(f);
            self.exports.get_string(index | 0x8000).set_value(name);
        }
    }

    pub fn nothing(sc: &ScCallContext) {
        sc.log("Doing nothing as requested. Oh, wait...");
    }
}
